ui/src/components/navigation/GlobalSearchModal.tsx

```diff
@@ -0,0 +1,347 @@
+import { Box, Fade, InputAdornment, Modal, TextField, Typography } from '@mui/material';
+
+import SongSearchResult from './search-results/SongSearchResult';
+import { useNavigate } from 'react-router-dom';
+import SetlistSearchResult from './search-results/SetlistSearchResult';
+import { SongSearchModalProps } from '#/types/song.types';
+import { SetlistSearchModalProps } from '#/types/setlist.types';
+import SearchIcon from '@mui/icons-material/Search';
+
+const RadioCard = (props: any) => {
+  const { children, isSelected, onClick } = props;
+
+  return (
+    <Box
+      sx={{
+        cursor: 'pointer',
+        border: 1,
+        borderRadius: '5px',
+        borderColor: isSelected ? '#6750A4' : '#938F99',
+        backgroundColor: isSelected ? '#6750A4' : '#2B2930',
+        fontWeight: isSelected ? '500' : 'normal',
+        px: '0.6rem',
+        py: '0.3rem',
+      }}
+      onClick={onClick}
+    >
+      <Typography variant="body2" fontWeight={isSelected ? '500' : 'normal'} color="#CAC4D0">
+        {children}
+      </Typography>
+    </Box>
+  );
+};
+
+type GlobalSearchModalProps = {
+  isOpen: boolean;
+  onClose: () => void;
+  allSongs: SongSearchModalProps[];
+  allSetlists: SetlistSearchModalProps[];
+};
+
+const GlobalSearchModal = (props: GlobalSearchModalProps) => {
+  const { isOpen, onClose, allSongs, allSetlists } = props;
+
+  const radioFilters = ['Songs', 'Setlists'];
+  const [radioFilter, setRadioFilter] = useState('Songs');
+
+  const [loading, setLoading] = useState(false);
+  const [filteredSongs, setFilteredSongs] = useState<SongSearchModalProps[]>([]);
+  const [filteredSetlists, setFilteredSetlists] = useState<SetlistSearchModalProps[]>([]);
+
+  const [searchString, setSearchString] = useState('');
+  const filterKeyword = useMemo(() => searchString.trim().toLowerCase(), [searchString]);
+
+  useEffect(() => {
+    setSearchString('');
+  }, [onClose]);
+
+  const filterSongs = () => {
+    if (filterKeyword.length < 2 || allSongs.length === 0) return [];
+
+    setLoading(true);
+    const filteredSongs = allSongs.filter((song) => {
+      const songTitle = song.title.toLowerCase();
+
+      if (songTitle.includes(filterKeyword)) {
+        return true;
+      }
+
+      return false;
+    });
+    setLoading(false);
+    return filteredSongs;
+  };
+
+  const filterSetlists = () => {
+    if (filterKeyword.length < 2 || allSetlists.length === 0) return [];
+
+    setLoading(true);
+    const filteredSetlists = allSetlists.filter((setlist) => {
+      const setlistName = setlist.name.toLowerCase();
+
+      if (setlistName.includes(filterKeyword)) {
+        return true;
+      }
+
+      return false;
+    });
+    setLoading(false);
+    return filteredSetlists;
+  };
+
+  const memoizedFilteredSongs = useMemo(() => {
+    return filterSongs();
+  }, [filterKeyword, allSongs]);
+
+  const memoizedFilteredSetlists = useMemo(() => {
+    return filterSetlists();
+  }, [filterKeyword, allSetlists]);
```

Comment:
having a variable that is not used inside useMemo, useCallback etc in the dependency array is generally bad practice. It's better if you just move the whole function inside the useMemo.